### PR TITLE
path restriction to same road 

### PR DESCRIPTION
**Pull request by andrewn6**
#44 

@ethndotsh

### CODE REVIEW

**Review by dabreegster — COMMENTED** [review]
Thanks for the PR! This is a good start, but there's a bit more we need. If you could open this PR against a non-main branch in your fork and allow me to push commits to it, I could chip away at it (but I'm traveling next few weeks and no guarantees)

- [ ] If you `cd examples./ serve_locally.sh` to build it, there's one build break -- in `set_area_mode`, need to set the new option to false
- [ ] Update `user_guide.md` near the `setRouteConfig` section
- [ ] Check the penalty -- is multiplying by 2 OK for this? It'd be nice to try this out a bit and maybe have screenshots / a manual sanity check this works as expected
- [ ] Add a checkbox to `examples` and plumb through in `lib.js`. (The code / API there is pretty clunky; I'm using a different TS + Svelte library in my own project. But still useful to control all the options from this demo)

File: route-snapper/src/lib.rs [hunk 1]
```diff
@@ -1024,10 +1025,15 @@ impl Router {
         // back
         // TODO Seems fast enough, but we could cache and build this up incrementally
         let mut avoid = HashSet::new();
+        let mut prev_edge_name = None;
+
         if self.config.avoid_doubling_back {
             for entry in prev_path {
                 if let PathEntry::Edge(e) = entry {
                     avoid.insert(e.0);
+                    if self.config.match_last_road {
```
Comment (dabreegster) [inline route-snapper/src/lib.rs:1034]: The path could be quite long and this potentially makes a bunch of unnecessary clones. Maybe we could do something like `prev_path.last()` or avoid the clone and do `map.edge(e).name.as_ref()` or something

File: route-snapper/src/lib.rs [hunk 2]
```diff
@@ -1039,7 +1045,7 @@ impl Router {
             node1,
             |i| i == node2,
             |(_, _, dir_edge)| {
-                let penalty = if avoid.contains(&dir_edge.0) {
+                let penalty = if avoid.contains(&dir_edge.0) || Some(self.map.edge(dir_edge.0).name.clone()) != prev_edge_name {
```
Comment (dabreegster) [inline route-snapper/src/lib.rs]: Same here -- even when the option is disabled, this'll clone every single edge name for the comparison. Maybe useful to split out a separate check for `match_last_road` and do the check inside there

**Comment by andrewn6** [comment]
Sounds good! i can work on this over the next week

File: route-snapper/src/lib.rs (hunk 1 above)
Comment (andrewn6) [reply route-snapper/src/lib.rs:1034]: fixed

https://github.com/dabreegster/route_snapper/pull/45/commits/bc42bf982426d1a7c3adc9857184238609f2b1fb

**Comment by andrewn6** [comment]
Ill be closing this #46

**Pull request closed without merging**